### PR TITLE
[Fixes #1744] Handling error when mongo is not available

### PR DESCRIPTION
**Pull request by vcpatil26**
https://thyssenkruppelevatorams.tpondemand.com/entity/1744

### CODE REVIEW

File: lib/mongo-stream.js [hunk 1]
```diff
@@ -25,8 +27,19 @@ MongoStream.prototype._write = function (chunk, encoding, done) {
 
     dao.writeLog(data, function (err) {
       if (err) {
-        console.error('fh-bunyan failed to write log to mongodb', err);
-        console.error((err && err.stack) ? err.stack : 'no stack available');
+        if (err.hasOwnProperty('jse_cause') &&
+          err.jse_cause.jse_shortmsg == 'failed to connect to mongodb' &&
+          !locked) {                    //if the error is caused by the unavailability of mongodb
+          console.error('MongoDB Connection error', err);
```
Comment (MikeyBurkman) [inline lib/mongo-stream.js]: You're still printing out the exception every time, except just with a different message when not locked. You want to print out the regular (`fh-bunyan failed to write log...`) message the first time, and swallow everything after that.

As for the locking, I think A) need to change the name, as we're not really locking anything, and B) we should just be swallowing everything after the first error. In the case of Node Gateway, we'd still see a nasty stacktrace every minute or so, which isn't good. 

Perhaps we could set the flag on an error, and reset the flag if we ever log without error? For Node Gateway, we'd never successfully log, so the flag would never reset, and thus we'd never see the error message again. In the other apps though, we'd see the error if the database was up, but later went down.

Comment (vcpatil26) [inline lib/mongo-stream.js]: @MikeyBurkman Updated the PR with the suggestions.

File: lib/mongo-stream.js [hunk 2]
```diff
@@ -25,8 +27,18 @@ MongoStream.prototype._write = function (chunk, encoding, done) {
 
     dao.writeLog(data, function (err) {
       if (err) {
-        console.error('fh-bunyan failed to write log to mongodb', err);
-        console.error((err && err.stack) ? err.stack : 'no stack available');
+        if (isMongoConnectionError(err) && !flag) {                    //if the error is caused by the unavailability of mongodb
+          console.error('fh-bunyan failed to write log to mongodb', err);
```
Comment (MikeyBurkman) [inline lib/mongo-stream.js]: You're still logging on every code branch. This means that you're logging on every error. You need to have a branch of the code where you're not logging.

File: lib/mongo-stream.js [hunk 3]
```diff
@@ -36,4 +48,15 @@ MongoStream.prototype._write = function (chunk, encoding, done) {
   }
 };
 
+function isMongoConnectionError(err) {
+  if (err.hasOwnProperty('jse_cause')) {
```
Comment (MikeyBurkman) [inline lib/mongo-stream.js]: Could simplify this: 

``` js
return (err.jse_cause && 
  err.jse_cause.jse_shortmsg === 'failed to connect to mongodb');
```

File: lib/mongo-stream.js [hunk 4]
```diff
@@ -2,9 +2,11 @@
 
 var WritableStream = require('stream').Writable
   , util = require('util')
-  , dao = require('./dao/bunyan-logs');
+  , dao = require('./dao/bunyan-logs'),
+  flag = false,
```
Comment (evanshortiss) [inline lib/mongo-stream.js]: I'm nitpicking, but could the existing style be used, or change the existing style of commas entirely 👍

File: lib/mongo-stream.js [hunk 5]
```diff
@@ -25,8 +27,20 @@ MongoStream.prototype._write = function (chunk, encoding, done) {
 
     dao.writeLog(data, function (err) {
       if (err) {
-        console.error('fh-bunyan failed to write log to mongodb', err);
-        console.error((err && err.stack) ? err.stack : 'no stack available');
+        if (isMongoConnectionError(err)) {                    //if the error is caused by the unavailability of mongodb
+          if (!flag) {
```
Comment (evanshortiss) [inline lib/mongo-stream.js:31]: Could this be moved to the above if to avoid nesting, like so?

``` js
if (isMongoConnectionError(err) && !flag)
```

Comment (MikeyBurkman) [inline lib/mongo-stream.js:31]: @evanshortiss That's is actually how he originally had it :) The problem is that we need 3 separate code branches of code to cover the possibilities:

1) Not a mongo connection error, log it
2) Mongo connection error, and we haven't seen one in a minute, so log it
3) Mongo connection error, but we have already logged one in the past minute, so ignore it.

File: lib/mongo-stream.js (hunk 4 above)
Comment (vcpatil26) [inline lib/mongo-stream.js]: @evanshortiss Using the existing styles now. :+1:

File: lib/mongo-stream.js (hunk 5 above)
Comment (evanshortiss) [inline lib/mongo-stream.js:31]: Sounds good to me 👍

**Pull request merged**